agents-audit/src/main/java/org/apache/ranger/audit/destination/ElasticSearchAuditDestination.java

```diff
@@ -0,0 +1,188 @@
+        super.stop();
+        logStatus();
+    }
+
+    @Override
+    public boolean log(Collection<AuditEventBase> events) {
+        boolean ret = false;
+		try {
+			logStatusIfRequired();
+			addTotalCount(events.size());
+
+            if (client == null) {
+				connect();
+				if (client == null) {
+					// Solr is still not initialized. So need return error
+					addDeferredCount(events.size());
+					return ret;
+				}
+			}
+
+			final Collection<Map<String, Object>> docs = new ArrayList<Map<String, Object>>();
+			for (AuditEventBase event : events) {
+				AuthzAuditEvent authzEvent = (AuthzAuditEvent) event;
+				docs.add(toDoc(authzEvent));
+			}
+			try {
+                IndexRequest indexRequest = new IndexRequest("posts").id("1").source(docs.toArray());
+                IndexResponse response = client.index(indexRequest, RequestOptions.DEFAULT);
+
+				if (response.status().getStatus() >= 400) {
+					addFailedCount(events.size());
+					logFailedEvent(events, response.toString());
+				} else {
+					addSuccessCount(events.size());
+					ret = true;
+				}
+			} catch (Exception ex) {
+				addFailedCount(events.size());
+				logFailedEvent(events, ex);
+			}
+		} catch (Throwable t) {
+			addDeferredCount(events.size());
+			logError("Error sending message to Solr", t);
+		}
+        return ret;
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see org.apache.ranger.audit.provider.AuditProvider#flush()
+     */
+    @Override
+    public void flush() {
+
+    }
+
+    public boolean isAsync() {
+        return true;
+    }
+
+    synchronized void connect() {
+        RestHighLevelClient me = client;
+        if (me == null) {
+            synchronized (ElasticSearchAuditDestination.class) {
+                me = client;
+                if (client == null) {
+                    String urls = MiscUtil.getStringProperty(props, propPrefix + "." + ELASTICSEARCH_URLS);
+                    String protocol = getStringProperty(props, propPrefix + "." + ELASTICSEARCH_PROTOCOL, "http");
+                    int port = MiscUtil.getIntProperty(props, propPrefix + "." + ELASTICSEARCH_PORT, 8080);
+                    if (urls != null) {
+                        urls = urls.trim();
+                    }
+                    if (urls != null && urls.equalsIgnoreCase("NONE")) {
+                        urls = null;
+                    }
+
+                    try {
+                        me = client = new RestHighLevelClient(
+                                RestClient.builder(
+                                        MiscUtil.toArray(urls, ",").stream()
+                                                .map(x -> new HttpHost(x, port, protocol))
+                                                .<HttpHost>toArray(i -> new HttpHost[i])
+                                ));
+                        ;
+                    } catch (Throwable t) {
+                        LOG.fatal("Can't connect to Solr server. urls=" + urls, t);
```

Comment:
Based on the logic, every time log is called and if the connection is not established, connect() is called. And if Elastic Search is not available, then we will continuously keep logging fatal. Since plugins run within process of the service, it will be bombarded with fatal logs. I feel, we should limit the number of logging. Probably keeping track of how many times we log per interval. E.g. we can have static variable which says last logged time and have a configurable what should be interval to log. If the interval > (currentTime - lastLogTime ), then log... Just a suggestion